File: app/integrations/http_client.py
from __future__ import annotations

import os
import time
from dataclasses import dataclass
from typing import Any, Dict, Optional

import httpx


@dataclass(frozen=True)
class HttpRetryPolicy:
    max_attempts: int = 3
    base_backoff_s: float = 0.2
    max_backoff_s: float = 2.0
    retry_on_status: tuple[int, ...] = (429, 500, 502, 503, 504)

    def backoff(self, attempt_idx: int) -> float:
        # attempt_idx starts at 0
        t = min(self.max_backoff_s, self.base_backoff_s * (2**attempt_idx))
        return t


_DEFAULT_TIMEOUT = httpx.Timeout(connect=2.0, read=6.0, write=6.0, pool=6.0)

# DEAD (currently unused): unused constant
DEFAULT_HEADERS: Dict[str, str] = {"User-Agent": "skylos-demo/0.1"}


def get_httpx_client(
    *,
    base_url: Optional[str] = None,
    timeout: httpx.Timeout = _DEFAULT_TIMEOUT,
) -> httpx.AsyncClient:
    verify_ssl = os.getenv("HTTP_VERIFY_SSL", "true").lower() != "false"

    headers = {
        "User-Agent": os.getenv("APP_USER_AGENT", "skylos-demo/0.1"),
        "Accept": "application/json",
    }

    return httpx.AsyncClient(
        base_url=base_url or "",
        timeout=timeout,
        headers=headers,
        verify=verify_ssl,
        follow_redirects=True,
    )


async def request_json(
    client: httpx.AsyncClient,
    method: str,
    url: str,
    *,
    json: Optional[Dict[str, Any]] = None,
    headers: Optional[Dict[str, str]] = None,
    retry: Optional[HttpRetryPolicy] = None,
) -> Dict[str, Any]:
    policy = retry or HttpRetryPolicy()

    last_exc: Optional[Exception] = None
    for attempt in range(policy.max_attempts):
        try:
            resp = await client.request(method, url, json=json, headers=headers)
            if resp.status_code in policy.retry_on_status:
                time.sleep(policy.backoff(attempt))
                continue
            resp.raise_for_status()
            data = resp.json()
            if isinstance(data, dict):
                return data
            return {"_": data}
        except Exception as e:
            last_exc = e
            time.sleep(policy.backoff(attempt))

    # if we got here, all attempts failed
    raise RuntimeError(f"request_json failed after {policy.max_attempts} attempts") from last_exc


# DEAD (currently unused): dead helper for text responses, never called by other module
async def request_text(
    client: httpx.AsyncClient,
    method: str,
    url: str,
    *,
    data: Optional[Dict[str, Any]] = None,
) -> str:
    resp = await client.request(method, url, data=data)
    resp.raise_for_status()
    return resp.text

File: web/src/integrations/httpClient.ts
// web/src/integrations/httpClient.ts
import http from "http";

export interface HttpRetryPolicy {
  maxAttempts: number;
  baseBackoffMs: number;
  maxBackoffMs: number;
  retryOnStatus: number[];
}

const defaultPolicy: HttpRetryPolicy = {
  maxAttempts: 3,
  baseBackoffMs: 200,
  maxBackoffMs: 2000,
  retryOnStatus: [429, 500, 502, 503, 504],
};

// UNUSED (demo): unused constant
export const DEFAULT_HEADERS: Record<string, string> = {
  "User-Agent": "skylos-demo/0.1",
};

// NOTE: re-exported in index.ts but never actually consumed anywhere
export function getHttpClient(): typeof http { // UNUSED (demo)
  return http;
}

export async function requestJson(
  url: string,
  options: { method?: string; body?: any; retry?: Partial<HttpRetryPolicy> } = {}
): Promise<any> {
  const policy = { ...defaultPolicy, ...options.retry };
  let lastError: Error | null = null;

  for (let attempt = 0; attempt < policy.maxAttempts; attempt++) {
    try {
      const resp = await fetch(url, {
        method: options.method || "GET",
        headers: { "Content-Type": "application/json" },
        body: options.body ? JSON.stringify(options.body) : undefined,
      });
      if (policy.retryOnStatus.includes(resp.status)) {
        await new Promise((r) => setTimeout(r, policy.baseBackoffMs * 2 ** attempt));
        continue;
      }
      return await resp.json();
    } catch (e) {
      lastError = e as Error;
      await new Promise((r) => setTimeout(r, policy.baseBackoffMs * 2 ** attempt));
    }
  }
  throw new Error(`requestJson failed after ${policy.maxAttempts} attempts: ${lastError?.message}`);
}

// UNUSED (demo): dead helper for text responses, never called
export async function requestText(
  url: string,
  options: { method?: string; body?: any } = {}
): Promise<string> {
  const resp = await fetch(url, {
    method: options.method || "GET",
    body: options.body ? JSON.stringify(options.body) : undefined,
  });
  return await resp.text();
}
